### PR TITLE
test(agw): Add state converter test for SPGW UE context

### PR DESCRIPTION
**Pull request by hcgatewood**
## Summary

#9603

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp [hunk 1]
```diff
@@ -50,7 +51,33 @@ TEST_F(SPGWStateConverterTest, TestSPGWStateConversion) {
   }
 }
 
-// TODO add a state conversion test for UE context
+TEST_F(SPGWStateConverterTest, TestUEContextConversion) {
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:62]: Let's add an additional unit test that expects the content of each `SpgwUeContext.S11BearerContext` against `s_plus_p_gw_eps_bearer_context_information_t` struct inside `state_teid_ht` hashtable to be the same

content to match:
  - Imsi_t imsi;
  - imsi64_t imsi64; 
  - int8_t imsi_unauthenticated_indicator;
  - char msisdn; 
  - teid_t mme_teid_S11; 
  - ip_address_t mme_ip_address_S11; 
  - teid_t s_gw_teid_S11_S4; 
  - ip_address_t s_gw_ip_address_S11_S4;  
  - ecgi_t last_known_cell_Id;

**Comment by hcgatewood** [comment]
@ardzoht now I'm getting a ld error... but I'm importing things properly best I can tell. Can you take a look and let me know if anything sticks out?

```
/usr/bin/ld: test/spgw_task/CMakeFiles/spgw_state_converter_test.dir/test_spgw_state_converter.cpp.o: in function `magma::lte::SPGWStateConverterTest_TestUEContextConversion_Test::TestBody()':
/home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:100: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:101: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:102: undefined reference to `imsi_to_imsi64(imsi_s const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:103: undefined reference to `imsi_to_imsi64(imsi_s const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:117: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:118: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:119: undefined reference to `imsi_to_imsi64(imsi_s const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:120: undefined reference to `imsi_to_imsi64(imsi_s const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
/usr/bin/ld: test/spgw_task/CMakeFiles/spgw_state_converter_test.dir/test_spgw_state_converter.cpp.o:/home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: more undefined references to `ip_address_to_bstring(ip_address_t const*)' follow
/usr/bin/ld: test/spgw_task/CMakeFiles/spgw_state_converter_test.dir/test_spgw_state_converter.cpp.o: in function `magma::lte::SPGWStateConverterTest_TestUEContextConversion_Test::TestBody()':
/home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:147: undefined reference to `ecgi_to_bytes(ecgi_s const*, char*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:148: undefined reference to `ecgi_to_bytes(ecgi_s const*, char*)'
/usr/bin/ld: test/spgw_task/libSPGW_TASK_TEST_LIB.a(state_creators.cpp.o): in function `magma::make_bearer_context(unsigned long, unsigned int)':
/home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/state_creators.cpp:67: undefined reference to `bstring_to_ip_address(tagbstring*, ip_address_t*)'
/usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/state_creators.cpp:71: undefined reference to `bstring_to_ip_address(tagbstring*, ip_address_t*)'
collect2: error: ld returned 1 exit status
[4/4] Linking CXX executable test/spgw_task/spgw_procedures_test
ninja: build stopped: subcommand failed.
ninja: build stopped: subcommand failed.
make: *** [Makefile:215: test_oai] Error 1
```

**Comment by ardzoht** [comment]
> @ardzoht now I'm getting a ld error... but I'm importing things properly best I can tell. Can you take a look and let me know if anything sticks out?
> 
> ```
> /usr/bin/ld: test/spgw_task/CMakeFiles/spgw_state_converter_test.dir/test_spgw_state_converter.cpp.o: in function `magma::lte::SPGWStateConverterTest_TestUEContextConversion_Test::TestBody()':
> /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:100: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:101: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:102: undefined reference to `imsi_to_imsi64(imsi_s const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:103: undefined reference to `imsi_to_imsi64(imsi_s const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:117: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:118: undefined reference to `imsi_string_to_3gpp_imsi(Imsi_t const*, imsi_s*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:119: undefined reference to `imsi_to_imsi64(imsi_s const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:120: undefined reference to `imsi_to_imsi64(imsi_s const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: undefined reference to `ip_address_to_bstring(ip_address_t const*)'
> /usr/bin/ld: test/spgw_task/CMakeFiles/spgw_state_converter_test.dir/test_spgw_state_converter.cpp.o:/home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:133: more undefined references to `ip_address_to_bstring(ip_address_t const*)' follow
> /usr/bin/ld: test/spgw_task/CMakeFiles/spgw_state_converter_test.dir/test_spgw_state_converter.cpp.o: in function `magma::lte::SPGWStateConverterTest_TestUEContextConversion_Test::TestBody()':
> /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:147: undefined reference to `ecgi_to_bytes(ecgi_s const*, char*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp:148: undefined reference to `ecgi_to_bytes(ecgi_s const*, char*)'
> /usr/bin/ld: test/spgw_task/libSPGW_TASK_TEST_LIB.a(state_creators.cpp.o): in function `magma::make_bearer_context(unsigned long, unsigned int)':
> /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/state_creators.cpp:67: undefined reference to `bstring_to_ip_address(tagbstring*, ip_address_t*)'
> /usr/bin/ld: /home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/state_creators.cpp:71: undefined reference to `bstring_to_ip_address(tagbstring*, ip_address_t*)'
> collect2: error: ld returned 1 exit status
> [4/4] Linking CXX executable test/spgw_task/spgw_procedures_test
> ninja: build stopped: subcommand failed.
> ninja: build stopped: subcommand failed.
> make: *** [Makefile:215: test_oai] Error 1
> ```

Removing line 15 (#include "common_types.h") on `state_creators.cpp` should fix this issue

**Comment by ardzoht** [comment]
@hcgatewood looks like there's merge conflicts on the PR, can you please rebase to fix those?

**Review by ardzoht — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp [hunk 2]
```diff
@@ -10,12 +10,18 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include <lte/gateway/c/core/oai/tasks/sgw/spgw_state_manager.h>
 #include <string.h>
 #include <gtest/gtest.h>
+#include <cstdio>
+#include <cstdlib>
 
+#include "ie_to_bytes.h"
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp]: can we enclose this (and the other plain C included files) on `extern "C"` directive?

**Pull request merged**